analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs

```diff
@@ -18,161 +18,139 @@
 
-            return methodCalledAsStatic
-                ? memberAccess.GetLocation()
-                : memberAccess.OperatorToken.CreateLocation(invocation);
-        }
+    private static Location GetReportLocation(InvocationExpressionSyntax invocation, bool methodCalledAsStatic) =>
+        methodCalledAsStatic is false && invocation.Expression is MemberAccessExpressionSyntax memberAccess
+            ? memberAccess.OperatorToken.CreateLocation(invocation)
+            : invocation.Expression.GetLocation();
 
-        private static ITypeSymbol GetElementType(InvocationExpressionSyntax invocation, IMethodSymbol methodSymbol,
-            SemanticModel semanticModel)
-        {
-            ExpressionSyntax collection;
-            if (methodSymbol.MethodKind == MethodKind.Ordinary)
+    private static ITypeSymbol GetElementType(InvocationExpressionSyntax invocation, IMethodSymbol methodSymbol, SemanticModel semanticModel)
+    {
+        return CollectionExpression() is { } collection
+            ? semanticModel.GetTypeInfo(collection).Type switch
             {
-                if (!invocation.ArgumentList.Arguments.Any())
-                {
-                    return null;
-                }
-                collection = invocation.ArgumentList.Arguments.First().Expression;
+                INamedTypeSymbol { TypeArguments: { Length: 1 } typeArguments } => typeArguments.First(),
+                IArrayTypeSymbol { Rank: 1 } arrayType => arrayType.ElementType,    // casting is necessary for multidimensional arrays
+                _ => null
             }
-            else
-            {
-                if (!(invocation.Expression is MemberAccessExpressionSyntax memberAccess))
+            : null;
+
+        ExpressionSyntax CollectionExpression() =>
+            methodSymbol.MethodKind is MethodKind.ReducedExtension
+                ? invocation.Expression switch
                 {
-                    return null;
+                    MemberAccessExpressionSyntax { Expression: { } memberAccessExpression } => memberAccessExpression,
+                    MemberBindingExpressionSyntax => invocation.GetParentConditionalAccessExpression().Expression,
+                    _ => null
                 }
-                collection = memberAccess.Expression;
-            }
-
-            var typeInfo = semanticModel.GetTypeInfo(collection);
-            if (typeInfo.Type is INamedTypeSymbol collectionType &&
-                collectionType.TypeArguments.Length == 1)
-            {
-                return collectionType.TypeArguments.First();
-            }
-
-            if (typeInfo.Type is IArrayTypeSymbol arrayType &&
-                arrayType.Rank == 1) // casting is necessary for multidimensional arrays
-            {
-                return arrayType.ElementType;
-            }
-
-            return null;
-        }
+                : invocation.ArgumentList.Arguments.FirstOrDefault()?.Expression;
```

Comment:
When we have `GetLeftSideOfDot` from Roslyn (see https://github.com/SonarSource/sonar-dotnet/issues/6863), it should be used to refactor this.